### ISSUE TITLE
High CPU Usage

### ISSUE THREAD

**Issue by skaldebane** [alainm23/planify#569]
### Description:
CPU usage reports 100% or higher using `top` and 2.5% using `gotop`.

### To Reproduce:
Just opening Planner and leaving it idle in the foreground. When it's on the background, it goes away.

### Expected behavior:
It shouldn't throttle my CPU. I have a very old laptop though I can run heavy things (e.g. Android Studio) but my fan stays calm when it's just idle and not building anything. While for Planner, the fans run their fastest possible speed (hence a lot of noise) to cool down the CPU...

This is weird, because Planner, unlike the heavy Android Studio, has to be a very lightweight application, all while performing all the scheduling and background tasks. Android Studio, on the other hand, still does a bunch of periodic background checks, file refreshes, auto-saving and much more when idle, though it feels as if it's "not even there", with almost zero impact on the performance.

I hope there's a way to fix this. I suspect it's related to the background checks and database operations, but you know best, and I hope you find a way to fix it.

### Screenshots:
On `top`:
![image](https://user-images.githubusercontent.com/35589610/96632524-c0804100-130f-11eb-803a-8b7d8d5d0f91.png)

On `gotop`:
![image](https://user-images.githubusercontent.com/35589610/96632461-a6466300-130f-11eb-9070-bf1346b249f9.png)

### System details:
 - OS & DE: Ubuntu 20.04; GNOME.
 - Version 2.5.4 (flatpak).
 - CPU: Intel Core i5 Gen 2.

**Comment by nsg** [comment]
Less noticeably for me on my system (newer computer), but I'm affected to. At my 5 years old work laptop it consumes closer to 25% of one core and is more noticeable.

![Screenshot from 2020-10-21 11-45-17](https://user-images.githubusercontent.com/16388/96703278-f286b700-1392-11eb-8ecf-393f9b02dbd4.png)

If I strace the process this is what's going on ... not sure if that's relevant.

![Screenshot from 2020-10-21 11-46-24](https://user-images.githubusercontent.com/16388/96703402-1813c080-1393-11eb-88dc-fa3044c3c23e.png)

The above system is running Ubuntu 20.04, my work computer is running 18.04. Both using the flatpak.

Possible related to #361

**Comment by nsg** [comment]
I investigated a little deeper yesterday and the file handle in question is an unix socket that talks with dbus, if that helps. This can of course be something unrelated. The POLLING event appears several times/second.

**Comment by skaldebane** [comment]
Yeah, it appears as if some infinite loop (intentional or unintentional) runs at its highest possible speed which leads to this high CPU usage... I don't know the code, but I wonder if we could put that repetitive task in a timer instead (if my hypothesis is true)

**Comment by manikk** [comment]
I too have the same behaviour. The CPU usage for Planner touches easily around 20% to 30%.

**Comment by igordsm** [comment]
Happens to me too. Ranges between 35% and 50%.

**Comment by Starcr0ss** [comment]
Same problem on my computer using Manjaro with KDE. Minimizing the window reduces the CPU usage, but keeping it open and using another window cranks up the CPU.

**Comment by skaldebane** [comment]
Small note: I've recently tried elementaryOS through a USB, and I noticed this issue with few system apps like the App Center, which does basically the same thing when idle.

**Comment by thetazero** [comment]
About 20% for me

**Comment by nsg** [comment]
> /.../ I noticed this issue with few system apps like the App Center, which does basically the same thing when idle.

I have considered it to be a bug in an Elementary framework, the question is which one so we can submit a bug.

**Comment by skaldebane** [comment]
Regarding the elementary App Center, they've apparently fixed this problem recently: [https://github.com/elementary/appcenter/issues/1160](https://github.com/elementary/appcenter/issues/1160).

This has been fixed in June and I tried the system myself in October. I might have been using an older version of App Center packaged with the system, as I didn't let it upgrade itself (it was booted directly from USB, just to try it).

Looking at the solution they found, it's not related to any frameworks or anything like that, it's just a screenshot cleaning code that was never stopping.

In the case of Planner, I guess it has something to do with the database operations or the auto syncing with Todoist.

**Comment by Skaldebane** [comment]
Hey again, I made a small search in elementary's GitHub profile and apparently many of their apps suffer from this.

![image](https://user-images.githubusercontent.com/35589610/100796940-ae6ee380-3421-11eb-8c80-4f0783854208.png)

Link: https://github.com/search?q=org%3Aelementary+high+cpu&type=issues

Any clues??

**Comment by srulikuk** [comment]
I am having the same issue. Flatpak planner v2.5.7 on ubuntu 18.04
Let me know if you need any logs.

**Comment by Hedronmx** [comment]
Having over 8% of cpu usage

![image](https://user-images.githubusercontent.com/26935602/103783426-90cb0600-4fed-11eb-96b5-b43f37064382.png)

**Comment by skaldebane** [comment]
Hi again! I just switched to elementaryOS (after my HDD died) and this issue is still there (in the Debian version, same as Flatpak).

**Comment by dtorres-sf** [comment]
I am experiencing this issue too, built from source - Ubuntu 18.04. Also I am seeing non-stop polling and retries (snippet from strace):

```recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 11) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 13) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 13) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 14) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 13) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 12) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 12) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 12) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 13) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 13) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 13) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 13) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 13) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 14) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 13) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 13) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 14) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 13) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}, {fd=20, events=POLLIN}, {fd=21, events=POLLIN}], 6, 14) = 0 (Timeout)
recvmsg(11, {msg_namelen=0}, 0)         = -1 EAGAIN (Resource temporarily unavailable)
```

**Comment by skaldebane** [comment]
Yeah, I suspect that's some sort of database issue or syncing code that's problematic there. That's my biggest complaint about Planner's impact on performance - if I leave the app window open. Of course the only workaround is to close it when you're done, and as closing it [**does not actually quit it**](https://ubuntu.com/blog/quit), it restores its state quickly.

**Comment by dtorres-sf** [comment]
@Skaldebane thank you for the workaround! I was always just minimizing to ensure I could use the global hot key to add a task from anywhere, but as you mentioned the app is still running and the global hot key still works.

**Comment by dmlls** [comment]
After quite some time checking the code, here go some thoughts:

Whenever Inbox, Today, or a project is clicked, a new `ListView` and `BoardView` are instanciated, e.g. in Inbox:

https://github.com/alainm23/planner/blob/59ef061c0f137fcc11732cd0e333513ecbaf1eb3/src/Views/Inbox.vala#L124-L128

Then, the `ListView`/`BoardView` adds the sections/boards, e.g. in `ListView` (same in `BoardView`):

https://github.com/alainm23/planner/blob/59ef061c0f137fcc11732cd0e333513ecbaf1eb3/src/Widgets/ListView.vala#L85-L99

This method instanciates `SectionRow`s / `BoardColumn`s. Each of these `SectionRow`s/ `BoardColumn`s refers to all the tasks in a section/column. Following up, each `SectionRow`/ `BoardColumn` loads **all the tasks** both completed and not completed, no matter whether the "Show Completed" switch is active or not, e.g. in `ListView`:

https://github.com/alainm23/planner/blob/59ef061c0f137fcc11732cd0e333513ecbaf1eb3/src/Widgets/SectionRow.vala#L371-L372

This instanciates an `ItemRow` for each item (task):

https://github.com/alainm23/planner/blob/59ef061c0f137fcc11732cd0e333513ecbaf1eb3/src/Widgets/SectionRow.vala#L1211-L1220

The code of [`ItemRow`](https://github.com/alainm23/planner/blob/master/src/Widgets/ItemRow.vala) is pretty complex (almost 2000 lines), adding a bunch of events and transitions to each of these items (tasks). In views such as Inbox it's fairly easy to end up with hundreds of tasks over time, meaning that every time you click on this view, all these tasks will be loaded regardless of whether they are completed or not, as already mentioned. This can be easily checked by printing a debug message in the `construct` of `ItemRow`:

![Screenshot_20210227_134632](https://user-images.githubusercontent.com/22967053/109387534-3b7ee700-7902-11eb-826f-692a3ad5849c.png)

So I guess it should be seriously considered to add some sort of lazy load mechanism so that completed tasks are only loaded when requested by the user, and even then probably it would be reasonable to load them progressively, not all at once.

This, of course, still doesn't solve the underlying problem which is that the code of `ItemRow` perhaps could be optimized somehow, so that even when hundreds of tasks are loaded, this doesn't result in an excessive CPU usage.

I might have got it all wrong, but I think evidence is quite strong. In any case, let's see what @alainm23 thinks, he should know better than anyone.

**Comment by alainm23** [comment]
@dmlls I'll consider your suggestions. I did some tests with `Widgets.ItemRow` creating many elements, commenting code, etc. But the performance remains the same. 

> Whenever Inbox, Today, or a project is clicked, a new ListView and BoardView are instanciated, e.g. in Inbox:

In previous versions, each Project view was stored in the cache every time it was initialized for the first time, this saved the load but loaded the CPU usage a lot, currently each project is initialized and cleaned at the time of selection, making use of the Low CPU and almost imperceptible to the user.

> This method instanciates SectionRows / BoardColumns. Each of these SectionRows/ BoardColumns refers to all the tasks in a section/column. Following up, each SectionRow/ BoardColumn loads all the tasks both completed and not completed, no matter whether the "Show Completed" switch is active or not, e.g. in ListView:

It's a good observation.

**Comment by dmlls** [comment]
Thanks Alain. Another thing to notice is that all these `ItemRow`s are instanciated both for `ListView` and `BoardView` (so I believe the same rows are instanciated twice) regardless of which view is active.

I've tried commenting out `add_all_items ()` and `add_completed_items ()` in `BoardColumn`, and moving `add_completed_items ()` in `SectionRow` (line [912](https://github.com/alainm23/planner/blob/59ef061c0f137fcc11732cd0e333513ecbaf1eb3/src/Widgets/SectionRow.vala#L921)):

```vala
Planner.database.project_show_completed.connect ((project) => {
    if (project.id == section.project_id) {
        if (project.show_completed == 1) {
            add_completed_items ();  // moved here
            completed_revealer.reveal_child = true;
        } else {
            completed_revealer.reveal_child = false;
        }
    }
});
```
so it only adds completed items when the switch is toggled. CPU usage didn't go over 20%, and then went back to around 12%. So I guess there's that. However, once "Show Completed"  is activated, CPU shoots up to 80% (I have around 180 tasks), and deactivating it doesn't bring it down. Furthermore, when activating it again, all the `ItemRows` are instanciated once again.

**Comment by dmlls** [comment]
Thanks Alain. Another thing to notice is that all these `ItemRow`s are instanciated both for `ListView` and `BoardView` (so I believe the same rows are instanciated twice) regardless of which view is active.

I've tried commenting out `add_all_items ()` and `add_completed_items ()` in `BoardColumn`, and moving `add_completed_items ()` in `SectionRow` (line [912](https://github.com/alainm23/planner/blob/59ef061c0f137fcc11732cd0e333513ecbaf1eb3/src/Widgets/SectionRow.vala#L921)):

```vala
Planner.database.project_show_completed.connect ((project) => {
    if (project.id == section.project_id) {
        if (project.show_completed == 1) {
            add_completed_items ();  // moved here
            completed_revealer.reveal_child = true;
        } else {
            completed_revealer.reveal_child = false;
        }
    }
});
```
so it only adds completed items when the switch is toggled. CPU usage didn't go over 20%, and then went back to around 12%. So I guess there's that. However, once "Show Completed"  is activated, CPU shoots up to 80% (I have around 180 tasks).

**Comment by skaldebane** [comment]
@alainm23 regarding the reloading of tasks, it lags for me (using an HDD) for lists that contain between 30~50 tasks (up to 1 second of freezing). It is perceptible for me, and I guess this applies to all HDD users. On SSD I guess it's gonna be blazing fast.

**Comment by kbdlnx** [comment]
Hey, guys so I have found a solution to the issue. I am not sure if it will work for everyone but what I did was to delete all my planner data by unlinking my account with Todoist and anything else, clearing all the setup and basically starting from scratch, and then the high CPU usage is no more. I hope this helps something!

**Comment by seguinj** [comment]
Thanks kbdlnx - resetting the application's data via Settings > General > Reset all did the trick for me. I've seen a huge performance increase since.

**Comment by AGI-chandler** [comment]
Still seeing this, I think, in the flatpak version 3.0.8.  I have an old laptop with 4 thread Intel i5-3210M.  Still my idle load average was 0.17, 0.20, 0.26 without Planner running.  With Planner open but hidden/not being used, it is constantly always using 15% CPU and load has gone up to 0.82, 0.59, 0.46 within about 15 mins.  If the underlying cause of this bug was actually addressed/fixed in the 3.0 releases, I have to ask: what on earth does it need all those CPU cycles for?

Otherwise I guess this bug still exists.  I've never used Planner before but it looked promising, so I gave it a try.  I therefore didn't have any old config files/data lying around that might be triggering this bug.  I didn't just spend 3 hours creating all my projects and todo lists in a fresh installation, to just delete them all and try again.  Seems to me the same thing will just happen?  There has to be a better workaround at least, or a fix to this bug once and for all.

**Comment by thedrint** [comment]
Ubuntu 22.04, Planify freshly installed from Flatpak.
High CPU usage (up to 90% on my laptop) when enabling **"autostart on logon"** option. In fact, app freezes after this change in app settings. All other app settings changed successfully, but only autostart on logon won't works properly - after terminate and restart app - this checkbox always unchecked.

I've add autostart of Planify manually via ubuntu autostarted apps system settings. WIll try to check this workaround, maybe it will helps.
UPD: Yes, in my case it works. No high CPU usage (near 10-12% as usual) if autostart option in app unchecked and app started on logon manally via system settings. Don't know what was reason of this strange high CPU usage, i'm not good at linux.
